src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -562,59 +593,41 @@ private bool FiltersMatch(RouteValueDictionary explicitValues)
                 if (IsRoutePartNonEmpty(kvp.Value))
                 {
                     var parameter = _pattern.GetParameter(kvp.Key);
-                    if (parameter == null && !context.AcceptedValues.ContainsKey(kvp.Key))
+                    if (parameter == null && !acceptedValues.ContainsKey(kvp.Key))
                     {
                         combinedValues.Add(kvp.Key, kvp.Value);
                     }
                 }
             }
         }
 
-        [DebuggerDisplay("{DebuggerToString(),nq}")]
-        private struct TemplateBindingContext
+        private static KeyValuePair<string, object>[] AssignSlots(RoutePattern pattern, KeyValuePair<string, object>[] filters)
         {
-            private readonly RouteValueDictionary _defaults;
-            private readonly RouteValueDictionary _acceptedValues;
-
-            public TemplateBindingContext(RouteValueDictionary defaults)
-            {
-                _defaults = defaults;
+            var slots = new KeyValuePair<string, object>[pattern.Parameters.Count + filters.Length];
 
-                _acceptedValues = new RouteValueDictionary();
-            }
-
-            public RouteValueDictionary AcceptedValues
+            for (var i = 0; i < pattern.Parameters.Count; i++)
             {
-                get { return _acceptedValues; }
+                slots[i] = new KeyValuePair<string, object>(pattern.Parameters[i].Name, null);
             }
 
-            public void Accept(string key, object value)
+            for (var i = 0; i < filters.Length; i++)
             {
-                if (!_acceptedValues.ContainsKey(key))
-                {
-                    _acceptedValues.Add(key, value);
-                }
+                slots[i + pattern.Parameters.Count] = new KeyValuePair<string, object>(filters[i].Key, null);
             }
 
-            public void AcceptDefault(string key)
-            {
-                Debug.Assert(!_acceptedValues.ContainsKey(key));
+            return slots;
+        }
 
-                if (_defaults != null && _defaults.TryGetValue(key, out var value))
-                {
-                    _acceptedValues.Add(key, value);
-                }
-            }
+        [DebuggerDisplay("explicit null")]
+        private class SentinullValue
```

Comment:
I lolled IRL

Reply:
There are comments where this is used but it would be good to have a sentence explaining what this is for here